### PR TITLE
pH Probe Calibration

### PR DESCRIPTION
**Pull request by je-foster**
Addresses #325. When the user is calibrating the pH probe:

- First the user is asked whether there will be 1, 2, or 3 calibration points.
- After the calibration points are submitted, a query is sent to the probe asking how many calibration points are active, and the probe's response is displayed to the user. This must be dismissed by the user pressing a key--it does not time out.
- The device remains in "calibration mode" for the duration of this process until the user dismisses the response. (Previously the device was not in calibration mode during the waits between entering the midpoint and the lowpoint and between entering the lowpoint and the highpoint.)

### CODE REVIEW

File: src/Devices/PHProbe.cpp [hunk 1]
```diff
@@ -100,27 +136,36 @@ void PHProbe::setHighpointCalibration(float highpoint) {
...
   char buffer[16];
   snprintf_P(buffer, sizeof(buffer), (PGM_P)F("Cal,mid,%i.%03i\r"), (int)midpoint, (int)(midpoint * 1000 + 0.5) % 1000);
   Serial1.print(buffer);  // send that string to the Atlas Scientific product
+#if defined(ARDUINO_CI_COMPILATION_MOCKS)
+  this->calibrationPoints = 1;
+#endif
   serial(F("PHProbe::setMidpointCalibration(%i.%03i)"), (int)midpoint, (int)(midpoint * 1000) % 1000);
 }
 
 #if defined(ARDUINO_CI_COMPILATION_MOCKS)
```
Comment (jgfoster) [inline src/Devices/PHProbe.cpp:165]: I see that you are setting the calibrationPoints earlier; I don't have the ability to search for senders so don't know how this method is used.

File: src/UIState/NumberCollectorState.h [hunk 2]
```diff
@@ -84,4 +87,13 @@ class TestIntNumCollectorState : public TestNumCollectorState {
   }
 };
 
+class TestNoCurrentNumCollectorState : public TestNumCollectorState {
```
Comment (jgfoster) [inline src/UIState/NumberCollectorState.h:90]: When I read the name of this class I thought that there was no number collector; perhaps the class name could be `TestNumCollectorStateWithNoCurrent` (but it didn't take much to figure out what is intended).

File: src/UIState/Wait.cpp [hunk 3]
```diff
@@ -3,6 +3,13 @@
 #include "MainMenu.h"
 #include "TC_util.h"
 
+/**
```
Comment (jgfoster) [inline src/UIState/Wait.cpp:6]: Nice documentation!

File: src/UIState/Wait.cpp [hunk 4]
```diff
@@ -14,8 +21,27 @@ Wait::Wait(TankController *tc, uint16_t msDelay, UIState *nextState) : UIState(t
...
+ */
+Wait::~Wait() {
+  if (nextState) {
+    delete nextState;
+    nextState = nullptr;
+  }
+}
+
+bool Wait::isInCalibration() {
+  if (nextState) {
```
Comment (jgfoster) [inline src/UIState/Wait.cpp:36]: This could be "simplified" (or at least fewer lines):
return nextState ? nextState->isInCalibration() : false;

**Review by jgfoster — CHANGES_REQUESTED** [review]

File: src/Devices/PHProbe.cpp [hunk 5]
```diff
@@ -38,21 +44,47 @@ PHProbe::PHProbe() {
...
+  snprintf_P(buffer, sizeof(buffer), (PGM_P)F("?CAL,%i\r"), this->calibrationPoints);
+  GODMODE()->serialPort[1].dataIn = buffer;  // the queue of data waiting to be read
+#endif
+  strscpy_P(calibrationResponse, F("Requesting..."), sizeof(calibrationResponse));
+}
+
+void PHProbe::getCalibration(char *buffer, int size) {
+  // for example "2" or "Requesting..."
+  strscpy(buffer, calibrationResponse, size);
+#if defined(ARDUINO_CI_COMPILATION_MOCKS)
```
[the remark this answers is not in the document]
Comment (je-foster) [reply src/Devices/PHProbe.cpp:66]: As written:

1. sendCalibrationRequest() populates the serial port but does not simulate the serial event.
2. getCalibration() returns the placeholder "Requesting..." string, simulating that the EZO pH probe has not yet replied the the request, but then triggers the serial event of receiving the message causing it to load the response into calibrationResponse.
3. A second getCalibration() returns the simulated response...and then triggers another serial event??

I don't like this anymore. I'll remove this mock part and trigger serialEvent1 in the test instead.

File: src/Devices/PHProbe.cpp [hunk 6]
```diff
@@ -6,6 +6,12 @@
 #include "Devices/Serial_TC.h"
 #include "TC_util.h"
 
+#if defined(ARDUINO_CI_COMPILATION_MOCKS)
```
Comment (jgfoster) [inline src/Devices/PHProbe.cpp:9]: Not needed here any more?

File: src/Devices/PHProbe.cpp [hunk 7]
```diff
@@ -72,9 +85,13 @@ void PHProbe::serialEvent1() {
         value = string.toFloat();
       } else if (string[0] == '?') {  // answer to a previous query
         serial(F("PHProbe serialEvent1: \"%s\""), string.c_str());
-        if (string.length() > 7 && string.substring(0, 7) == "?SLOPE,") {
+        if (string.length() > 7 && memcmp_P(string.c_str(), F("?SLOPE,"), 7) == 0) {
           // for example "?SLOPE,16.1,100.0"
-          strscpy(slopeResponse, string.c_str(), sizeof(slopeResponse));  // Flawfinder: ignore
+          strscpy(slopeResponse, string.c_str() + 7, sizeof(slopeResponse));  // Flawfinder: ignore
```
Comment (jgfoster) [inline src/Devices/PHProbe.cpp:90]: Try without Flawfinder to see if it helps (and on line 94).

File: test/PHProbeTest.cpp [hunk 8]
```diff
@@ -21,14 +21,52 @@ unittest(serialEvent1) {
...
   assertEqual(7.125, pPHProbe->getPh());
+  assertEqual("99.7,100.3,-0.89", pPHProbe->getSlopeResponse());
+}
+
+unittest(clearCalibration) {
+  GodmodeState *state = GODMODE();
+  state->reset();
+  assertEqual("", state->serialPort[1].dataOut);
+  PHProbe::instance()->clearCalibration();
+  assertEqual("Cal,clear\r", state->serialPort[1].dataOut);
```
Comment (jgfoster) [inline test/PHProbeTest.cpp:43]: Case?

File: test/PHProbeTest.cpp [hunk 9]
```diff
@@ -21,14 +21,52 @@ unittest(serialEvent1) {
...
+}
+
+unittest(sendCalibrationRequest) {
+  GodmodeState *state = GODMODE();
+  state->reset();
+  assertEqual("", state->serialPort[1].dataOut);
+  PHProbe::instance()->sendCalibrationRequest();
+  assertEqual("CAL,?\r", state->serialPort[1].dataOut);
+  char buffer[20];
+  PHProbe::instance()->getCalibration(buffer, sizeof(buffer));
```
Comment (jgfoster) [inline test/PHProbeTest.cpp:53]: Need buffer?

File: test/PHProbeTest.cpp [hunk 10]
```diff
@@ -82,9 +120,12 @@ unittest(setHighpointCalibration) {
 unittest(sendSlopeRequest) {
   GodmodeState *state = GODMODE();
   state->reset();
-  state->serialPort[0].dataOut = "";
+  assertEqual("", state->serialPort[1].dataOut);
   PHProbe::instance()->sendSlopeRequest();
-  assertEqual("SLOPE,?\r", GODMODE()->serialPort[1].dataOut);
+  assertEqual("SLOPE,?\r", state->serialPort[1].dataOut);
+  char buffer[20];
+  PHProbe::instance()->getSlope(buffer, sizeof(buffer));
```
Comment (jgfoster) [inline test/PHProbeTest.cpp:127]: buffer?

**Review by jgfoster — CHANGES_REQUESTED** [review]

File: src/UIState/Wait.cpp (hunk 3 above)
Comment (je-foster) [reply src/UIState/Wait.cpp:6]: Well I typed "/**" and VSCode did the rest...

**Review by jgfoster — APPROVED** [review]

**Comment by jgfoster** [comment]
Fix #325.

**Pull request merged**